### DISCUSSION TITLE
Azure AD Authentication, Next-Auth JWT and .NET Core Web API

### DISCUSSION THREAD

**Discussion by verb0se** [nextauthjs/next-auth#5996] [Help]
### Question 💬

I've set up a Next.js app to authenticate to Azure AD using Next-Auth, and this is successfully returning a JWT token. I'd like to use this token for authentication in a .NET Core Web API project but I'm getting a 401 Unauthorized error with the WWW-Authenticate header returning Bearer error="invalid_token". I'm using the same secret for the value of both the NEXTAUTH_SECRET and for configuring the IssuerSigningKey property in the API project's startup, which seems to be working properly as I'm not getting a "signature is invalid" error.  I've validated that the JWT is well formed and contains no errors at https://jwt.io.  If anyone knows a resource that shows the proper configuration for this or if they can spot a configuration error in my code it's appreciated.

[...nextauth].ts:

```
import NextAuth, { NextAuthOptions } from "next-auth"
import AzureADProvider from "next-auth/providers/azure-ad";

export const authOptions: NextAuthOptions = {
  providers: [
    AzureADProvider({
      clientId: process.env.AZURE_AD_CLIENT_ID!,
      clientSecret: process.env.AZURE_AD_CLIENT_SECRET!,
      tenantId: process.env.AZURE_AD_TENANT_ID,
      authorization: {
        params: {
          scope:
            `api://${process.env.AZURE_AD_CLIENT_ID}/access_as_user`,
        },
      }
    }),
  ],
  theme: {
    colorScheme: "light",
  },
  
  callbacks: {
    async jwt({ token, user, account, profile, isNewUser }) {
      if (account) {
        token.access_token = account.access_token;
      }
      return token;
    },
    async session({ session, token }) {
      return { ...session };
    },
  },
}

export default NextAuth(authOptions)
```

weather.ts:
```
import { getToken } from 'next-auth/jwt'
import type { NextApiRequest, NextApiResponse } from "next"

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const token = await getToken({ req, secret: process.env.NEXTAUTH_SECRET }) // get JWT token from request

  try {
    return new Promise(async (resolve, reject) => {
        const apiUrl = `${process.env.API_BASE_URL}/weatherForecast/get`;

        const response = await fetch(apiUrl, { // example URL
            method: "GET",
            credentials: "include",
            headers: {
              Authorization: `Bearer ${token?.access_token}`,  // <-- add token to request
              ContentType: "application/json",
            }
          });
          const json = await response.json();
          res.send(JSON.stringify(json, null, 2));               
    });
} catch (e) {
    var error = e;
  }
}
```

Program.cs:
```
using System.Text;

using Microsoft.AspNetCore.Authentication.JwtBearer;
using Microsoft.Identity.Web;
using Microsoft.IdentityModel.Logging;
using Microsoft.IdentityModel.Tokens;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
.AddJwtBearer(o =>
{
    o.RequireHttpsMetadata = false;
    //o.Audience = "00000003-0000-0000-c000-000000000000";
    o.Audience = "http://localhost:3000";
    o.Authority = builder.Configuration["AzureAd:Instance"];

    o.TokenValidationParameters = new TokenValidationParameters
    {
        IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(builder.Configuration["Jwt:Key"])), //Same as NEXTAUTH_SECRET
        ValidIssuer = o.Authority,
        ValidAudience = o.Audience,
        ClockSkew = TimeSpan.Zero 
    };
})
.AddMicrosoftIdentityWebApi(builder.Configuration.GetSection("AzureAd"), "AzureAd");

builder.Services.AddControllers();

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

builder.Services.AddCors(o => o.AddPolicy("default", builder =>
{
    builder.AllowAnyOrigin()
            .AllowAnyMethod()
            .AllowAnyHeader();
}));

var app = builder.Build();

IdentityModelEventSource.ShowPII = true;

if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
    app.UseDeveloperExceptionPage();
}

app.UseHttpsRedirection();

app.UseAuthentication();
app.UseAuthorization();

app.MapControllers();

app.Run();
```

### How to reproduce ☕️

1.  Use next-auth sample template and configure Azure AD authentication
2. Create .NET Core Web API and configure API protection via Azure AD authentication
3. Create typescript page that fetches from Web API, passing encoded JWT as Bearer token in Authentication header 

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by aarondewindt** [comment]
I'm running into a similar issue. In my case I've noticed that the requested custom scope doesn't appear in the JWT scp field. In my nextauth configuration I have:

```ts
...

AzureADProvider({
  clientId: process.env.AZURE_AD_CLIENT_ID ?? "",
  clientSecret: process.env.AZURE_AD_CLIENT_SECRET ?? "",
  tenantId: process.env.AZURE_AD_TENANT_ID,
  authorization: {
    params: {
      scope: "email GroupMember.Read.All openid profile User.Read api://xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/read"
    }
  }
}),

...
```

But the scp field in the JWT token only contains: "email GroupMember.Read.All openid profile User.Read". Is this also your case?

For context my backend is a django application with DRF and django-auth-adfs. I can login using the login pages of nextauth and django-auth-adfs. But I can't use the access_token I'm getting from nextauth to authorize API calls into the django app. I get a `"Invalid access token."` error msg back from the server. The only weird think I can see right now is the missing scope in the JWT token.

**Reply by aarondewindt** [reply]
Sovled thanks to a comment on a similar question.
https://github.com/nextauthjs/next-auth/discussions/3006#discussioncomment-2466512

TL;DR You need to leave the nextauth/AzureADProvider default scopes, so that nextauth can log the user in and only after the user has logged in can you request a token for the API using MSAL.

The code in the comment worked perfectly fine for me. The only change I had to make was replace the requested scope in the MSAL call.
